src/binutils/doc/cxxfilt.man
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man 5.01 (Pod::Simple 3.45)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "C++FILT 1"
.TH C++FILT 1 2026-06-03 binutils-2.16.91 "GNU Development Tools"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
cxxfilt \- Demangle C++ and Java symbols.
.SH SYNOPSIS
.IX Header "SYNOPSIS"
c++filt [\fB\-_\fR|\fB\-\-strip\-underscores\fR]
        [\fB\-j\fR|\fB\-\-java\fR]
        [\fB\-n\fR|\fB\-\-no\-strip\-underscores\fR]
        [\fB\-p\fR|\fB\-\-no\-params\fR]
        [\fB\-s\fR \fIformat\fR|\fB\-\-format=\fR\fIformat\fR]
        [\fB\-\-help\fR]  [\fB\-\-version\fR]  [\fIsymbol\fR...]
.SH DESCRIPTION
.IX Header "DESCRIPTION"
The C++ and Java languages provides function overloading, which means
that you can write many functions with the same name (providing each
takes parameters of different types).  All C++ and Java function names
are encoded into a low-level assembly label (this process is known as
\&\fImangling\fR). The \fBc++filt\fR
[1]
program does the inverse mapping: it decodes (\fIdemangles\fR) low-level
names into user-level names so that the linker can keep these overloaded
functions from clashing.
.PP
Every alphanumeric word (consisting of letters, digits, underscores,
dollars, or periods) seen in the input is a potential label.  If the
label decodes into a C++ name, the C++ name replaces the low-level
name in the output.
.PP
You can use \fBc++filt\fR to decipher individual symbols:
.PP
.Vb 1
\&        c++filt <symbol>
.Ve
.PP
If no \fIsymbol\fR arguments are given, \fBc++filt\fR reads symbol
names from the standard input and writes the demangled names to the
standard output.  All results are printed on the standard output.
.SH OPTIONS
.IX Header "OPTIONS"
.IP \fB\-_\fR 4
.IX Item "-_"
.PD 0
.IP \fB\-\-strip\-underscores\fR 4
.IX Item "--strip-underscores"
.PD
On some systems, both the C and C++ compilers put an underscore in front
of every name.  For example, the C name \f(CW\*(C`foo\*(C'\fR gets the low-level
name \f(CW\*(C`_foo\*(C'\fR.  This option removes the initial underscore.  Whether
\&\fBc++filt\fR removes the underscore by default is target dependent.
.IP \fB\-j\fR 4
.IX Item "-j"
.PD 0
.IP \fB\-\-java\fR 4
.IX Item "--java"
.PD
Prints demangled names using Java syntax.  The default is to use C++
syntax.
.IP \fB\-n\fR 4
.IX Item "-n"
.PD 0
.IP \fB\-\-no\-strip\-underscores\fR 4
.IX Item "--no-strip-underscores"
.PD
Do not remove the initial underscore.
.IP \fB\-p\fR 4
.IX Item "-p"
.PD 0
.IP \fB\-\-no\-params\fR 4
.IX Item "--no-params"
.PD
When demangling the name of a function, do not display the types of
the function's parameters.
.IP "\fB\-s\fR \fIformat\fR" 4
.IX Item "-s format"
.PD 0
.IP \fB\-\-format=\fR\fIformat\fR 4
.IX Item "--format=format"
.PD
\&\fBc++filt\fR can decode various methods of mangling, used by
different compilers.  The argument to this option selects which
method it uses:
.RS 4
.ie n .IP """auto""" 4
.el .IP \f(CWauto\fR 4
.IX Item "auto"
Automatic selection based on executable (the default method)
.ie n .IP """gnu""" 4
.el .IP \f(CWgnu\fR 4
.IX Item "gnu"
the one used by the GNU C++ compiler (g++)
.ie n .IP """lucid""" 4
.el .IP \f(CWlucid\fR 4
.IX Item "lucid"
the one used by the Lucid compiler (lcc)
.ie n .IP """arm""" 4
.el .IP \f(CWarm\fR 4
.IX Item "arm"
the one specified by the C++ Annotated Reference Manual
.ie n .IP """hp""" 4
.el .IP \f(CWhp\fR 4
.IX Item "hp"
the one used by the HP compiler (aCC)
.ie n .IP """edg""" 4
.el .IP \f(CWedg\fR 4
.IX Item "edg"
the one used by the EDG compiler
.ie n .IP """gnu\-v3""" 4
.el .IP \f(CWgnu\-v3\fR 4
.IX Item "gnu-v3"
the one used by the GNU C++ compiler (g++) with the V3 ABI.
.ie n .IP """java""" 4
.el .IP \f(CWjava\fR 4
.IX Item "java"
the one used by the GNU Java compiler (gcj)
.ie n .IP """gnat""" 4
.el .IP \f(CWgnat\fR 4
.IX Item "gnat"
the one used by the GNU Ada compiler (GNAT).
.RE
.RS 4
.RE
.IP \fB\-\-help\fR 4
.IX Item "--help"
Print a summary of the options to \fBc++filt\fR and exit.
.IP \fB\-\-version\fR 4
.IX Item "--version"
Print the version number of \fBc++filt\fR and exit.
.SH FOOTNOTES
.IX Header "FOOTNOTES"
.IP 1. 4
MS-DOS does not allow \f(CW\*(C`+\*(C'\fR characters in file names, so on
MS-DOS this program is named \fBCXXFILT\fR.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
the Info entries for \fIbinutils\fR.
.SH COPYRIGHT
.IX Header "COPYRIGHT"
Copyright (c) 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999,
2000, 2001, 2002, 2003, 2004, 2005 Free Software Foundation, Inc.
.PP
Permission is granted to copy, distribute and/or modify this document
under the terms of the GNU Free Documentation License, Version 1.1
or any later version published by the Free Software Foundation;
with no Invariant Sections, with no Front-Cover Texts, and with no
Back-Cover Texts.  A copy of the license is included in the
section entitled ``GNU Free Documentation License''.


ChangeLog
2026-05-24  Eric Gallager  <1388660+cooljeanius@users.noreply.github.com>

	Merge pull request #202 from zaidanfaiz/fix-cgen-guile3-conditional-defines

	Fix CGEN conditional definitions for Guile 3

2026-05-24  Eric Gallager  <egall@gwmail.gwu.edu>

	Testing cgen subdir Makefile targets

	The following Makefile targets work in the cgen subdir with --target=arm-apple-darwin:
	- html-local
	- sim-arch
	The following ones fail:
	- opcodes
	- sim-cpu
	- gas-test
	- sim-test

	Change-Id: I32144ef234c439c59e97d6b6bbfcf9da7bcf2764

2026-05-24  Eric Gallager  <1388660+cooljeanius@users.noreply.github.com>

	Merge pull request #203 from cooljeanius/dependabot/npm_and_yarn/src/texinfo/js/npm-07f3ad912e

	Bump the npm group across 1 directory with 2 updates

2026-05-24  dependabot[bot]  <49699333+dependabot[bot]@users.noreply.github.com>

	Bump the npm group across 1 directory with 2 updates

	Bumps the npm group with 2 updates in the /src/texinfo/js directory: [dompurify](https://github.com/cure53/DOMPurify) and [eslint](https://github.com/eslint/eslint).


	Updates `dompurify` from 3.4.2 to 3.4.3
	- [Release notes](https://github.com/cure53/DOMPurify/releases)
	- [Commits](https://github.com/cure53/DOMPurify/compare/3.4.2...3.4.3)

	Updates `eslint` from 10.3.0 to 10.4.0
	- [Release notes](https://github.com/eslint/eslint/releases)
	- [Commits](https://github.com/eslint/eslint/compare/v10.3.0...v10.4.0)

	---
	updated-dependencies:
	- dependency-name: dompurify
	  dependency-version: 3.4.3
	  dependency-type: direct:production
	  update-type: version-update:semver-patch
	  dependency-group: npm
	- dependency-name: eslint
	  dependency-version: 10.4.0
	  dependency-type: direct:development
	  update-type: version-update:semver-minor
	  dependency-group: npm
	...

2026-05-24  Eric Gallager  <1388660+cooljeanius@users.noreply.github.com>

	Merge pull request #201 from cooljeanius/dependabot/github_actions/actions-0f1baada16

	Bump the actions group across 1 directory with 4 updates

2026-05-23  dependabot[bot]  <49699333+dependabot[bot]@users.noreply.github.com>

	Bump the actions group across 1 directory with 4 updates

	Bumps the actions group with 4 updates in the / directory: [actions/checkout](https://github.com/actions/checkout), [mozilla-actions/sccache-action](https://github.com/mozilla-actions/sccache-action), [actions/upload-artifact](https://github.com/actions/upload-artifact) and [github/codeql-action](https://github.com/github/codeql-action).


	Updates `actions/checkout` from 4 to 6
	- [Release notes](https://github.com/actions/checkout/releases)
	- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
	- [Commits](https://github.com/actions/checkout/compare/v4...v6)

	Updates `mozilla-actions/sccache-action` from 0.0.9 to 0.0.10
	- [Release notes](https://github.com/mozilla-actions/sccache-action/releases)
	- [Commits](https://github.com/mozilla-actions/sccache-action/compare/v0.0.9...v0.0.10)

	Updates `actions/upload-artifact` from 4 to 7
	- [Release notes](https://github.com/actions/upload-artifact/releases)
	- [Commits](https://github.com/actions/upload-artifact/compare/v4...v7)

	Updates `github/codeql-action` from 4.35.3 to 4.35.5
	- [Release notes](https://github.com/github/codeql-action/releases)
	- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
	- [Commits](https://github.com/github/codeql-action/compare/v4.35.3...v4.35.5)

	---
	updated-dependencies:
	- dependency-name: actions/checkout
	  dependency-version: '6'
	  dependency-type: direct:production
	  update-type: version-update:semver-major
	  dependency-group: actions
	- dependency-name: actions/upload-artifact
	  dependency-version: '7'
	  dependency-type: direct:production
	  update-type: version-update:semver-major
	  dependency-group: actions
	- dependency-name: github/codeql-action
	  dependency-version: 4.35.4
	  dependency-type: direct:production
	  update-type: version-update:semver-patch
	  dependency-group: actions
	- dependency-name: mozilla-actions/sccache-action
	  dependency-version: 0.0.10
	  dependency-type: direct:production
	  update-type: version-update:semver-patch
	  dependency-group: actions
	...

2026-05-16  Codex  <codex@openai.com>

	Fix CGEN conditional definitions for Guile 3

2026-05-16  Eric Gallager  <1388660+cooljeanius@users.noreply.github.com>

	Merge pull request #200 from cooljeanius/dependabot/gradle/gradle-2978c64675

	Bump com.google.guava:guava from 33.5.0-jre to 33.6.0-jre in the gradle group

2026-05-16  dependabot[bot]  <49699333+dependabot[bot]@users.noreply.github.com>

	Bump com.google.guava:guava in the gradle group

	Bumps the gradle group with 1 update: [com.google.guava:guava](https://github.com/google/guava).


	Updates `com.google.guava:guava` from 33.5.0-jre to 33.6.0-jre
	- [Release notes](https://github.com/google/guava/releases)
	- [Commits](https://github.com/google/guava/commits)

	---
	updated-dependencies:
	- dependency-name: com.google.guava:guava
	  dependency-version: 33.6.0-jre
	  dependency-type: direct:production
	  update-type: version-update:semver-minor
	  dependency-group: gradle
	...

2026-05-12  Eric Gallager  <egall@gwmail.gwu.edu>

	shuffle workflows around a bit more

	Change-Id: I8a9136fcca0d7f0ed02238a70c0ef212155bb25b

2026-05-12  Eric Gallager  <1388660+cooljeanius@users.noreply.github.com>

	Merge pull request #199 from cooljeanius/macos_arm

	Add aarch64-apple-darwin support

2026-05-12  Eric Gallager  <egall@gwmail.gwu.edu>

	noticed things during review

	Change-Id: I3c14c7cdbb1963b16a9e8490a10b896e0a342e62

2026-05-12  Eric Gallager  <egall@gwmail.gwu.edu>

	Merge branch 'master' into macos_arm

	Change-Id: Ic0f04a827eb5c57bbd7de9e39342f69dc5895572

2026-05-12  Eric Gallager  <egall@gwmail.gwu.edu>

	a few minor workflow updates

	Change-Id: Ie69cce10eb05a518ad49389ab260d80080387185

2026-05-12  Eric Gallager  <egall@gwmail.gwu.edu>

	more frustrations, mig and (separately) gettext


src/binutils/cxxfilt.c
/* cxxfilt.c
 * Demangler for GNU C++ - main program
 * Copyright 1989, 1991, 1994, 1995, 1996, 1997, 1998, 1999,
 * 2000, 2001, 2002, 2003 Free Software Foundation, Inc.
 * Written by James Clark <jjc@jclark.uucp>
 * Rewritten by Fred Fish <fnf@cygnus.com> for ARM and Lucid demangling
 * Modified by Satish Pai <pai@apollo.hp.com> for HP demangling */
/*
This file is part of GCC.

GCC is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 2, or (at your option) any later
version.

GCC is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with GCC; see the file COPYING.  If not, write to the Free
Software Foundation, 51 Franklin Street - Fifth Floor, Boston, MA
02110-1301, USA.  */

#include "config.h"
#include "bfd.h"
#include "bucomm.h"
#include "libiberty.h"
#include "demangle.h"
#include "getopt.h"
#include "safe-ctype.h"

#include "sysdep.h"

static int flags = (DMGL_PARAMS | DMGL_ANSI | DMGL_VERBOSE);

static void demangle_it (char *);
static void usage (FILE *, int) ATTRIBUTE_NORETURN;
static void print_demangler_list (FILE *);

static void
demangle_it(char *mangled_name)
{
  char *result;

  /* For command line args, also try to demangle type encodings.  */
  result = cplus_demangle(mangled_name, (flags | DMGL_TYPES));
  if (result == NULL) {
      printf ("%s\n", mangled_name);
  } else {
      printf("%s\n", result);
      free(result);
  }
}

static void
print_demangler_list(FILE *stream)
{
  const struct demangler_engine *demangler;

  fprintf(stream, "{%s", libiberty_demanglers->demangling_style_name);

  for ((demangler = (libiberty_demanglers + 1));
       (demangler->demangling_style != unknown_demangling);
       ++demangler) {
	  fprintf(stream, ",%s", demangler->demangling_style_name);
  }

  fprintf(stream, "}");
}

/* */
static void
usage(FILE *stream, int status)
{
  fprintf(stream, "\
Usage: %s [-_] [-n] [--strip-underscores] [--no-strip-underscores]\n\
       [-p] [--no-params]\n",
	   program_name);

  fprintf(stream, "\
       [-s ");
  print_demangler_list(stream);
  fprintf(stream, "]\n");

  fprintf(stream, "\
       [--format ");
  print_demangler_list(stream);
  fprintf(stream, "]\n");

  fprintf(stream, "\
       [--help] [--version] [arg...]\n");
  xexit(status);
}

static char mbuffer[32767];

extern int strip_underscore;
int strip_underscore = 0;

static const struct option long_options[] = {
  { "strip-underscores", no_argument, 0, '_' },
  { "format", required_argument, 0, 's' },
  { "help", no_argument, 0, 'h' },
  { "no-params", no_argument, 0, 'p' },
  { "no-strip-underscores", no_argument, 0, 'n' },
  { "version", no_argument, 0, 'v' },
  { 0, no_argument, 0, 0 }
};

static const char *standard_symbol_characters(void);

static const char *hp_symbol_characters(void);

/* Return the string of non-alnum characters that may occur
 * as a valid symbol component, in the standard assembler symbol
 * syntax.  */

static const char *
standard_symbol_characters(void)
{
  return "_$.";
}


/* Return the string of non-alnum characters that may occur
 * as a valid symbol name component in an HP object file.
 *
 * Note that, since HP's compiler generates object code straight from
 * C++ source, without going through an assembler, its mangled
 * identifiers can use all sorts of characters that no assembler would
 * tolerate, so the alphabet this function creates is a little odd.
 * Here are some sample mangled identifiers offered by HP:
 *
 *	typeid*__XT24AddressIndExpClassMember_
 *	[Vftptr]key:__dt__32OrdinaryCompareIndExpClassMemberFv
 *	__ct__Q2_9Elf64_Dyn18{unnamed.union.#1}Fv
 *
 * This still seems really weird to me, since nowhere else in this
 * file is there anything to recognize curly brackets, parens, etc.
 * I have talked with Srikanth <srikanth@cup.hp.com>, and he assures me
 * this is right, but I still strongly suspect that there is a
 * misunderstanding here.
 *
 * If we decide that it is better for c++filt to use HP's assembler syntax
 * to scrape identifiers out of its input, here is the definition of
 * the symbol name syntax from the HP assembler manual:
 *
 *     Symbols are composed of uppercase and lowercase letters, decimal
 *     digits, dollar symbol, period (.), ampersand (&), pound sign(#) and
 *     underscore (_). A symbol can begin with a letter, digit underscore or
 *     dollar sign. If a symbol begins with a digit, it must contain a
 *     non-digit character.
 *
 * So have fun. */
static const char *
hp_symbol_characters(void)
{
  return "_$.<>#,*&[]:(){}";
}

extern int main(int, char **);

int
main(int argc, char **argv)
{
  char *result;
  int c;
  const char *valid_symbols;
  enum demangling_styles style = auto_demangling;

  program_name = argv[0];
  xmalloc_set_program_name (program_name);

  strip_underscore = TARGET_PREPENDS_UNDERSCORE;

  while ((c = getopt_long(argc, argv, "_nps:", long_options, (int *)0)) != EOF) {
      switch (c) {
        case '?':
          usage(stderr, 1);
#if !defined(__clang__)
          break;
#endif /* !__clang__ */
        case 'h':
          usage(stdout, 0);
        case 'n':
          strip_underscore = 0;
          break;
        case 'p':
          flags &= ~ DMGL_PARAMS;
          break;
        case 'v':
          print_version("c++filt");
          return (0);
        case '_':
          strip_underscore = 1;
          break;
        case 's':
          {
            style = cplus_demangle_name_to_style (optarg);
            if (style == unknown_demangling) {
              fprintf(stderr, "%s: unknown demangling style `%s'\n",
                      program_name, optarg);
              return (1);
            } else {
              cplus_demangle_set_style(style);
            }
          }
          break;
        default:
          break;
      } /* end "switch (c)" */
  } /* end while-loop */

  if (optind < argc) {
      for (; (optind < argc); optind++) {
        demangle_it(argv[optind]);
      }
  } else {
      switch (current_demangling_style) {
        case gnu_demangling:
        case lucid_demangling:
        case arm_demangling:
        case java_demangling:
        case edg_demangling:
        case gnat_demangling:
        case gnu_v3_demangling:
        case dlang_demangling:
        case auto_demangling:
          valid_symbols = standard_symbol_characters();
          break;
        case hp_demangling:
          valid_symbols = hp_symbol_characters();
          break;
        case no_demangling:
          valid_symbols = NULL; /* FIXME: is this correct? */
          break;
        case unknown_demangling: /* fall through */
        default:
          /* Folks should explicitly indicate the appropriate alphabet for
           * each demangling. Providing a default would allow the
           * question to go unconsidered. */
          fatal("Internal error: no symbol alphabet for current style");
      } /* end "switch (current_demangling_style)" */

      for (;;) {
        unsigned i = 0U;
        c = getchar();
        /* Try to read a label: */
        while ((c != EOF) && (ISALNUM(c) || strchr(valid_symbols, c))) {
          if (i >= (sizeof(mbuffer) - 1)) {
            break;
          }
          mbuffer[i++] = (char)c;
          c = getchar();
        }
        if (i > 0) {
          unsigned skip_first = 0;

          mbuffer[i] = 0;
          if ((mbuffer[0] == '.') || (mbuffer[0] == '$')) {
            ++skip_first;
          }
          if (strip_underscore && (mbuffer[skip_first] == '_')) {
            ++skip_first;
          }

          if (skip_first > i) {
            skip_first = i;
          }

          flags |= (int)style;
          result = cplus_demangle((mbuffer + skip_first), flags);
          if (result) {
            if (mbuffer[0] == '.') {
              putc('.', stdout);
            }
            fputs(result, stdout);
            free(result);
          } else {
            fputs(mbuffer, stdout);
          }

          fflush(stdout);
        }
        if (c == EOF) {
          break;
        }
        putchar(c);
        fflush(stdout);
      } /* end for-loop */
  }

  return(0);
}

/* EOF */
